src/Orleans/Serialization/TypeUtilities.cs

```diff
@@ -82,7 +82,7 @@ internal static bool IsOrleansShallowCopyable(this Type t)
 
         private static bool IsTypeFieldsShallowCopyable(TypeInfo typeInfo)
         {
-            return typeInfo.GetFields().All(f => !(f.FieldType.IsEquivalentTo(typeInfo)) && IsOrleansShallowCopyable(f.FieldType));
+            return typeInfo.GetFields().All(f => !f.FieldType.GetTypeInfo().IsEquivalentTo(typeInfo.AsType()) && IsOrleansShallowCopyable(f.FieldType));
```

Comment:
There's no need to revert original PR, as it purpose was in extracting allocation causing code into separate method. So it's OK as it is.